File: linkis-engineconn-plugins/shell/src/main/scala/org/apache/linkis/manager/engineplugin/shell/executor/ShellEngineConnExecutor.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.manager.engineplugin.shell.executor

import org.apache.linkis.common.utils.{Logging, Utils}
import org.apache.linkis.engineconn.computation.executor.execute.{
  ComputationExecutor,
  EngineExecutionContext
}
import org.apache.linkis.engineconn.core.EngineConnObject
import org.apache.linkis.governance.common.utils.GovernanceUtils
import org.apache.linkis.manager.common.entity.resource.{
  CommonNodeResource,
  LoadInstanceResource,
  NodeResource
}
import org.apache.linkis.manager.engineplugin.common.conf.EngineConnPluginConf
import org.apache.linkis.manager.engineplugin.common.util.NodeResourceUtils
import org.apache.linkis.manager.engineplugin.shell.common.ShellEngineConnPluginConst
import org.apache.linkis.manager.engineplugin.shell.exception.ShellCodeErrorException
import org.apache.linkis.manager.label.entity.Label
import org.apache.linkis.protocol.engine.JobProgressInfo
import org.apache.linkis.rpc.Sender
import org.apache.linkis.scheduler.executer.{
  ErrorExecuteResponse,
  ExecuteResponse,
  SuccessExecuteResponse
}

import org.apache.commons.io.IOUtils
import org.apache.commons.lang3.StringUtils

import java.io.{BufferedReader, File, FileReader, InputStreamReader, IOException}
import java.util
import java.util.concurrent.CountDownLatch
import java.util.concurrent.atomic.AtomicBoolean

import scala.collection.JavaConverters._
import scala.collection.mutable.ArrayBuffer

class ShellEngineConnExecutor(id: Int) extends ComputationExecutor with Logging {

  private var engineExecutionContext: EngineExecutionContext = _

  private val executorLabels: util.List[Label[_]] = new util.ArrayList[Label[_]]()

  private var process: Process = _

  private var extractor: YarnAppIdExtractor = _

  override def init(): Unit = {
    logger.info(s"Ready to change engine state!")
    super.init
  }

  override def executeCompletely(
      engineExecutionContext: EngineExecutionContext,
      code: String,
      completedLine: String
  ): ExecuteResponse = {
    val newcode = completedLine + code
    logger.debug("newcode is " + newcode)
    executeLine(engineExecutionContext, newcode)
  }

  override def executeLine(
      engineExecutionContext: EngineExecutionContext,
      code: String
  ): ExecuteResponse = {

    if (null != engineExecutionContext) {
      this.engineExecutionContext = engineExecutionContext
      logger.info("Shell executor reset new engineExecutionContext!")
    }

    var bufferedReader: BufferedReader = null
    var errorsReader: BufferedReader = null

    val completed = new AtomicBoolean(false)
    var errReaderThread: ReaderThread = null
    var inputReaderThread: ReaderThread = null

    try {
      engineExecutionContext.appendStdout(s"$getId >> ${code.trim}")

      val argsArr =
        if (
            engineExecutionContext.getTotalParagraph == 1 &&
            engineExecutionContext.getProperties != null &&
            engineExecutionContext.getProperties.containsKey(
              ShellEngineConnPluginConst.RUNTIME_ARGS_KEY
            )
        ) {
          Utils.tryCatch {
            val argsList = engineExecutionContext.getProperties
              .get(ShellEngineConnPluginConst.RUNTIME_ARGS_KEY)
              .asInstanceOf[util.ArrayList[String]]
            logger.info(
              "Will execute shell task with user-specified arguments: \'" + argsList
                .toArray(new Array[String](argsList.size()))
                .mkString("\' \'") + "\'"
            )
            argsList.toArray(new Array[String](argsList.size()))
          } { t =>
            logger.warn(
              "Cannot read user-input shell arguments. Will execute shell task without them.",
              t
            )
            null
          }
        } else {
          null
        }

      val workingDirectory =
        if (
            engineExecutionContext.getTotalParagraph == 1 &&
            engineExecutionContext.getProperties != null &&
            engineExecutionContext.getProperties.containsKey(
              ShellEngineConnPluginConst.SHELL_RUNTIME_WORKING_DIRECTORY
            )
        ) {
          Utils.tryCatch {
            val wdStr = engineExecutionContext.getProperties
              .get(ShellEngineConnPluginConst.SHELL_RUNTIME_WORKING_DIRECTORY)
              .asInstanceOf[String]
            if (isExecutePathExist(wdStr)) {
              logger.info(
                "Will execute shell task under user-specified working-directory: \'" + wdStr
              )
              wdStr
            } else {
              logger.warn(
                "User-specified working-directory: \'" + wdStr + "\' does not exist or user does not have access permission. " +
                  "Will execute shell task under default working-directory. Please contact the administrator!"
              )
              null
            }
          } { t =>
            logger.warn(
              "Cannot read user-input working-directory. Will execute shell task under default working-directory.",
              t
            )
            null
          }
        } else {
          null
        }

      val generatedCode = if (argsArr == null || argsArr.isEmpty) {
        generateRunCode(code)
      } else {
        generateRunCodeWithArgs(code, argsArr)
      }

      val processBuilder: ProcessBuilder = new ProcessBuilder(generatedCode: _*)
      if (StringUtils.isNotBlank(workingDirectory)) {
        processBuilder.directory(new File(workingDirectory))
      }

      processBuilder.redirectErrorStream(false)
      extractor = new YarnAppIdExtractor
      process = processBuilder.start()

      bufferedReader = new BufferedReader(new InputStreamReader(process.getInputStream))
      errorsReader = new BufferedReader(new InputStreamReader(process.getErrorStream))
      val counter: CountDownLatch = new CountDownLatch(2)
      inputReaderThread =
        new ReaderThread(engineExecutionContext, bufferedReader, extractor, true, counter)
      errReaderThread =
        new ReaderThread(engineExecutionContext, errorsReader, extractor, false, counter)

      inputReaderThread.start()
      errReaderThread.start()

      val exitCode = process.waitFor()
      counter.await()

      completed.set(true)

      if (exitCode != 0) {
        ErrorExecuteResponse("run shell failed", ShellCodeErrorException())
      } else SuccessExecuteResponse()
    } catch {
      case e: Exception =>
        logger.error("Execute shell code failed, reason:", e)
        ErrorExecuteResponse("run shell failed", e)
    } finally {
      if (null != errorsReader) {
        inputReaderThread.onDestroy()
      }
      if (null != inputReaderThread) {
        errReaderThread.onDestroy()
      }
      IOUtils.closeQuietly(bufferedReader)
      IOUtils.closeQuietly(errorsReader)
    }
  }

  private def isExecutePathExist(executePath: String): Boolean = {
    val etlHomeDir: File = new File(executePath)
    (etlHomeDir.exists() && etlHomeDir.isDirectory())
  }

  private def generateRunCode(code: String): Array[String] = {
    Array("sh", "-c", code)
  }

  private def generateRunCodeWithArgs(code: String, args: Array[String]): Array[String] = {
    Array("sh", "-c", "echo \"dummy " + args.mkString(" ") + "\" | xargs sh -c \'" + code + "\'")
  }

  override def getId(): String = Sender.getThisServiceInstance.getInstance + "_" + id

  override def getProgressInfo(taskID: String): Array[JobProgressInfo] = {
    val jobProgressInfo = new ArrayBuffer[JobProgressInfo]()
    if (null == this.engineExecutionContext) {
      return jobProgressInfo.toArray
    }
    if (0.0f == progress(taskID)) {
      jobProgressInfo += JobProgressInfo(engineExecutionContext.getJobId.getOrElse(""), 1, 1, 0, 0)
    } else {
      jobProgressInfo += JobProgressInfo(engineExecutionContext.getJobId.getOrElse(""), 1, 0, 0, 1)
    }
    jobProgressInfo.toArray
  }

  override def progress(taskID: String): Float = {
    if (null != this.engineExecutionContext) {
      this.engineExecutionContext.getCurrentParagraph / this.engineExecutionContext.getTotalParagraph
        .asInstanceOf[Float]
    } else {
      0.0f
    }
  }

  override def supportCallBackLogs(): Boolean = {
    // todo
    true
  }

  override def requestExpectedResource(expectedResource: NodeResource): NodeResource = {
    null
  }

  override def getCurrentNodeResource(): NodeResource = {
    val resource = new CommonNodeResource
    resource.setUsedResource(
      NodeResourceUtils
        .applyAsLoadInstanceResource(EngineConnObject.getEngineCreationContext.getOptions)
    )
    resource
  }

  override def getExecutorLabels(): util.List[Label[_]] = executorLabels

  override def setExecutorLabels(labels: util.List[Label[_]]): Unit = {
    if (null != labels) {
      executorLabels.clear()
      executorLabels.addAll(labels)
    }
  }

  override def killTask(taskID: String): Unit = {
    /*
      Kill sub-processes
     */
    val pid = getPid(process)
    GovernanceUtils.killProcess(String.valueOf(pid), s"kill task $taskID process", false)
    /*
      Kill yarn-applications
     */
    val yarnAppIds = extractor.getExtractedYarnAppIds()
    GovernanceUtils.killYarnJobApp(yarnAppIds)
    logger.info(s"Finished kill yarn app ids in the engine of (${getId()}).}")
    super.killTask(taskID)

  }

  private def getPid(process: Process): Int = {
    Utils.tryCatch {
      val clazz = Class.forName("java.lang.UNIXProcess");
      val field = clazz.getDeclaredField("pid");
      field.setAccessible(true);
      field.get(process).asInstanceOf[Int]
    } { t =>
      logger.warn("Failed to acquire pid for shell process")
      -1
    }
  }

  override def close(): Unit = {
    try {
      process.destroy()
    } catch {
      case e: Exception =>
        logger.error(s"kill process ${process.toString} failed ", e)
      case t: Throwable =>
        logger.error(s"kill process ${process.toString} failed ", t)
    }
    super.close()
  }

}

File: linkis-engineconn-plugins/shell/src/main/java/org/apache/linkis/manager/engineplugin/shell/executor/ShellEngineConnExecutor.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.manager.engineplugin.shell.executor;

import org.apache.linkis.common.utils.Utils;
import org.apache.linkis.engineconn.computation.executor.execute.ComputationExecutor;
import org.apache.linkis.engineconn.computation.executor.execute.EngineExecutionContext;
import org.apache.linkis.engineconn.core.EngineConnObject;
import org.apache.linkis.governance.common.utils.GovernanceUtils;
import org.apache.linkis.manager.common.entity.resource.CommonNodeResource;
import org.apache.linkis.manager.common.entity.resource.NodeResource;
import org.apache.linkis.manager.engineplugin.common.util.NodeResourceUtils;
import org.apache.linkis.manager.engineplugin.shell.common.ShellEngineConnPluginConst;
import org.apache.linkis.manager.engineplugin.shell.conf.ShellEngineConnConf;
import org.apache.linkis.manager.engineplugin.shell.exception.ShellCodeErrorException;
import org.apache.linkis.manager.label.entity.Label;
import org.apache.linkis.protocol.engine.JobProgressInfo;
import org.apache.linkis.rpc.Sender;
import org.apache.linkis.scheduler.executer.ErrorExecuteResponse;
import org.apache.linkis.scheduler.executer.ExecuteResponse;
import org.apache.linkis.scheduler.executer.SuccessExecuteResponse;

import org.apache.commons.lang3.StringUtils;

import java.io.BufferedReader;
import java.io.File;
import java.io.InputStreamReader;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicBoolean;

import scala.concurrent.ExecutionContextExecutorService;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ShellEngineConnExecutor extends ComputationExecutor {
  private static final Logger logger = LoggerFactory.getLogger(ShellEngineConnExecutor.class);

  private EngineExecutionContext engineExecutionContext;

  private List<Label<?>> executorLabels = new ArrayList<>();

  Map<String, ShellECTaskInfo> shellECTaskInfoCache = new ConcurrentHashMap<>();

  private int id;

  private Process process;

  private YarnAppIdExtractor extractor;

  public ShellEngineConnExecutor(int id) {
    super(ShellEngineConnConf.SHELL_ENGINECONN_OUTPUT_PRINT_LIMIT);
    this.id = id;
  }

  final ExecutionContextExecutorService logAsyncService =
      Utils.newCachedExecutionContext(
          ShellEngineConnConf.LOG_SERVICE_MAX_THREAD_SIZE, "ShelLogService-Thread-", true);

  @Override
  public void init() {
    logger.info("Ready to change engine state!");
    super.init();
  }

  @Override
  public ExecuteResponse executeCompletely(
      EngineExecutionContext engineExecutionContext, String code, String completedLine) {
    String newcode = completedLine + code;
    logger.debug("newcode is " + newcode);
    return executeLine(engineExecutionContext, newcode);
  }

  @Override
  public ExecuteResponse executeLine(EngineExecutionContext engineExecutionContext, String code) {
    if (engineExecutionContext != null) {
      this.engineExecutionContext = engineExecutionContext;
      logger.info("Shell executor reset new engineExecutionContext!");
    }

    if (engineExecutionContext.getJobId().isEmpty()) {
      return new ErrorExecuteResponse("taskID is null", null);
    }

    String taskId = engineExecutionContext.getJobId().get();
    BufferedReader bufferedReader = null;
    BufferedReader errorsReader = null;

    AtomicBoolean completed = new AtomicBoolean(false);
    ReaderThread errReaderThread = null;
    ReaderThread inputReaderThread = null;

    try {
      engineExecutionContext.appendStdout(getId() + " >> " + code.trim());

      String[] argsArr;
      if (engineExecutionContext.getTotalParagraph() == 1
          && engineExecutionContext.getProperties() != null
          && engineExecutionContext
              .getProperties()
              .containsKey(ShellEngineConnPluginConst.RUNTIME_ARGS_KEY)) {
        ArrayList<String> argsList =
            (ArrayList<String>)
                engineExecutionContext
                    .getProperties()
                    .get(ShellEngineConnPluginConst.RUNTIME_ARGS_KEY);
        argsArr = argsList.toArray(new String[argsList.size()]);
        logger.info(
            "Will execute shell task with user-specified arguments: '{}'",
            StringUtils.join(argsArr, "' '"));
      } else {
        argsArr = null;
      }

      String workingDirectory;
      if (engineExecutionContext.getTotalParagraph() == 1
          && engineExecutionContext.getProperties() != null
          && engineExecutionContext
              .getProperties()
              .containsKey(ShellEngineConnPluginConst.SHELL_RUNTIME_WORKING_DIRECTORY)) {
        String wdStr =
            (String)
                engineExecutionContext
                    .getProperties()
                    .get(ShellEngineConnPluginConst.SHELL_RUNTIME_WORKING_DIRECTORY);
        if (isExecutePathExist(wdStr)) {
          logger.info(
              "Will execute shell task under user-specified working-directory: '{}'", wdStr);
          workingDirectory = wdStr;
        } else {
          logger.warn(
              "User-specified working-directory: '{}' does not exist or user does not have access permission. Will execute shell task under default working-directory. Please contact the administrator!",
              wdStr);
          workingDirectory = null;
        }
      } else {
        workingDirectory = null;
      }

      String[] generatedCode =
          argsArr == null || argsArr.length == 0
              ? generateRunCode(code)
              : generateRunCodeWithArgs(code, argsArr);

      ProcessBuilder processBuilder = new ProcessBuilder(generatedCode);

      if (StringUtils.isNotBlank(workingDirectory)) {
        processBuilder.directory(new File(workingDirectory));
      }

      processBuilder.redirectErrorStream(false);
      extractor = new YarnAppIdExtractor();
      process = processBuilder.start();
      bufferedReader = new BufferedReader(new InputStreamReader(process.getInputStream()));
      errorsReader = new BufferedReader(new InputStreamReader(process.getErrorStream()));

      // add task id and task Info cache
      shellECTaskInfoCache.put(taskId, new ShellECTaskInfo(taskId, process, extractor));

      CountDownLatch counter = new CountDownLatch(2);
      inputReaderThread =
          new ReaderThread(engineExecutionContext, bufferedReader, extractor, true, counter);
      errReaderThread =
          new ReaderThread(engineExecutionContext, errorsReader, extractor, false, counter);

      logAsyncService.execute(inputReaderThread);
      logAsyncService.execute(errReaderThread);

      int exitCode = process.waitFor();
      counter.await();

      completed.set(true);

      if (exitCode != 0) {
        return new ErrorExecuteResponse("run shell failed", new ShellCodeErrorException());
      } else {
        return new SuccessExecuteResponse();
      }
    } catch (Exception e) {
      logger.error("Execute shell code failed, reason:", e);
      return new ErrorExecuteResponse("run shell failed", e);
    } finally {
      if (errorsReader != null) {
        errReaderThread.onDestroy();
      }
      if (inputReaderThread != null) {
        inputReaderThread.onDestroy();
      }
      shellECTaskInfoCache.remove(taskId);
    }
  }

  private boolean isExecutePathExist(String executePath) {
    File etlHomeDir = new File(executePath);
    return (etlHomeDir.exists() && etlHomeDir.isDirectory());
  }

  private String[] generateRunCode(String code) {
    return new String[] {"sh", "-c", code};
  }

  private String[] generateRunCodeWithArgs(String code, String[] args) {
    return new String[] {
      "sh",
      "-c",
      "echo \"dummy " + StringUtils.join(args, " ") + "\" | xargs sh -c \'" + code + "\'"
    };
  }

  @Override
  public String getId() {
    return Sender.getThisServiceInstance().getInstance() + "_" + id;
  }

  @Override
  public JobProgressInfo[] getProgressInfo(String taskID) {
    List<JobProgressInfo> jobProgressInfos = new ArrayList<>();
    if (this.engineExecutionContext == null) {
      return jobProgressInfos.toArray(new JobProgressInfo[0]);
    }

    String jobId =
        engineExecutionContext.getJobId().isDefined()
            ? engineExecutionContext.getJobId().get()
            : "";
    if (progress(taskID) == 0.0f) {
      jobProgressInfos.add(new JobProgressInfo(jobId, 1, 1, 0, 0));
    } else {
      jobProgressInfos.add(new JobProgressInfo(jobId, 1, 0, 0, 1));
    }
    return jobProgressInfos.toArray(new JobProgressInfo[0]);
  }

  @Override
  public float progress(String taskID) {
    if (this.engineExecutionContext != null) {
      return ((float) this.engineExecutionContext.getCurrentParagraph())
          / this.engineExecutionContext.getTotalParagraph();
    } else {
      return 0.0f;
    }
  }

  @Override
  public boolean supportCallBackLogs() {
    // todo
    return true;
  }

  @Override
  public NodeResource requestExpectedResource(NodeResource expectedResource) {
    return null;
  }

  @Override
  public NodeResource getCurrentNodeResource() {
    CommonNodeResource resource = new CommonNodeResource();
    resource.setUsedResource(
        NodeResourceUtils.applyAsLoadInstanceResource(
            EngineConnObject.getEngineCreationContext().getOptions()));
    return resource;
  }

  @Override
  public List<Label<?>> getExecutorLabels() {
    return executorLabels;
  }

  @Override
  public void setExecutorLabels(List<Label<?>> labels) {
    if (labels != null) {
      executorLabels.clear();
      executorLabels.addAll(labels);
    }
  }

  @Override
  public void killTask(String taskID) {

    /*
     Kill sub-processes
    */
    int pid = getPid(process);
    GovernanceUtils.killProcess(String.valueOf(pid), "kill task " + taskID + " process", false);

    /*
     Kill yarn-applications
    */
    List<String> yarnAppIds = extractor.getExtractedYarnAppIds();

    GovernanceUtils.killYarnJobApp(yarnAppIds);
    logger.info("Finished kill yarn app ids in the engine of ({})", getId());
    super.killTask(taskID);
  }

  int getPid(Process process) {
    try {
      Class<?> clazz = Class.forName("java.lang.UNIXProcess");
      Field field = clazz.getDeclaredField("pid");
      field.setAccessible(true);
      return field.getInt(process);
    } catch (Exception e) {
      logger.warn("Failed to acquire pid for shell process");
      return -1;
    }
  }

  @Override
  public void close() {
    try {
      process.destroy();
    } catch (Exception e) {
      logger.error("kill process " + process.toString() + " failed ", e);
    } catch (Throwable t) {
      logger.error("kill process " + process.toString() + " failed ", t);
    }
    super.close();
  }
}
